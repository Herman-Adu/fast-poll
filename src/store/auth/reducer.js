import { SIGN_IN, SIGN_OUT, LOADING } from './actions';

export const INITIAL_STATE = {
  loading: false,
  uid: '',
  isAnonymous: null,
  photoURL: '',
  // // some other properties from the user object that may be useful
  // email: '',
  // displayName: '',
};

export default (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case SIGN_IN:
      return {
        ...state,
        ...action.payload,
      };

    case SIGN_OUT:
      return INITIAL_STATE;

    case LOADING:
      return {
        ...state,
        loading: action.payload.isLoading,
      };

    default:
      return state;
  }
};
