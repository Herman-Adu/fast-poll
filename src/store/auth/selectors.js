import { createSelector } from 'reselect';

export const selectUid = state => state.auth.uid;
export const selectIsAnonymous = state => state.auth.isAnonymous;

export const selectAuthedState = createSelector(
  [selectUid, selectIsAnonymous],
  (uid, isAnonymous) => {
    // If uid doesn't exist in state, the user is not signed in.
    // A uid will exist if the user is signed in anonymously.
    // We'll consider anonymous users as unauthed for this variable.
    return !!(uid && !isAnonymous);
  },
);
