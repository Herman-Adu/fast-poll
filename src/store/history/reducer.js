import { LOADING, ADD_POLLS } from './actions';

export const INITIAL_STATE = {
  created: [],
  votedOn: [],
  loading: false,
};

export default (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case LOADING:
      return {
        ...state,
        loading: action.payload.isLoading,
      };

    case ADD_POLLS:
      return {
        ...state,
        [action.payload.type]: [
          ...state[action.payload.type],
          ...action.payload.polls,
        ],
      };

    default:
      return state;
  }
};
