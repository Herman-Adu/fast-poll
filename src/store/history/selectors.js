import { createSelector } from 'reselect';

import { sortPollsByProperty } from './utils';

export const selectCreatedPolls = state => state.history.created;
export const selectVotedOnPolls = state => state.history.votedOn;

export const selectSortedPolls = selector =>
  createSelector([selector], polls => {
    return sortPollsByProperty(polls, 'date');
  });
