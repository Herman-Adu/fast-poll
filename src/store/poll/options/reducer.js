import {
  ADD_OPTION,
  REMOVE_OPTION,
  UPDATE_OPTION,
  UPDATE_OPTION_ORDER,
  RESET_OPTIONS,
  UPDATE_NEWEST,
} from './actions';

export const INITIAL_STATE = {
  // [optionId]
  order: [],
  // options is a map with the optionId as the keys and
  // option data including results as the values
  data: {},
  newest: {
    id: null,
    text: '',
  },
};

export default (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case ADD_OPTION:
      return {
        ...state,
        order: [...state.order, action.payload.id],
        data: {
          ...state.data,
          [action.payload.id]: {
            id: action.payload.id,
            editing: false,
            votes: 0,
            ...action.payload.data,
          },
        },
      };

    case REMOVE_OPTION:
      const { [action.payload.id]: removedOption, ...newOptions } = state.data;

      return {
        ...state,
        data: newOptions,
        order: state.order.filter(id => id !== action.payload.id),
      };

    case UPDATE_OPTION:
      return {
        ...state,
        data: {
          ...state.data,
          [action.payload.id]: {
            ...state.data[action.payload.id],
            ...Object.keys(action.payload.data).reduce((aggr, curr) => {
              if (curr === 'votes') {
                // eslint-disable-next-line no-param-reassign
                aggr.votes =
                  action.payload.id in state.data &&
                  'votes' in state.data[action.payload.id]
                    ? state.data[action.payload.id].votes +
                      action.payload.data.votes
                    : action.payload.data.votes;
              } else {
                // eslint-disable-next-line no-param-reassign
                aggr[curr] = action.payload.data[curr];
              }
              return aggr;
            }, {}),
          },
        },
      };

    case UPDATE_OPTION_ORDER:
      return {
        ...state,
        order: [...action.payload.order],
      };

    case UPDATE_NEWEST:
      return {
        ...state,
        newest: {
          ...state.newest,
          ...action.payload.data,
        },
      };

    case RESET_OPTIONS:
      return INITIAL_STATE;

    default:
      return state;
  }
};
