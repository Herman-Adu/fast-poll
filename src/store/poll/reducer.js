import { combineReducers } from 'redux';

import optionsReducer from './options/reducer';
import {
  LOADING,
  UPDATE_TITLE,
  UPDATE_SELECTION,
  SET_SHOW_RESULTS,
  SET_CREATED_POLL,
  SET_OWNER,
  RESET_POLL,
} from './actions';

export const INITIAL_STATE = {
  loading: false,
  title: '',
  selection: '',
  showResults: false,
  created: '',
  createdBy: '',
};

const rootReducer = (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case LOADING:
      return {
        ...state,
        loading: action.payload.isLoading,
      };

    case UPDATE_TITLE:
      return {
        ...state,
        title: action.payload.title,
      };

    case UPDATE_SELECTION:
      return {
        ...state,
        selection: action.payload.selection,
      };

    case SET_SHOW_RESULTS:
      return {
        ...state,
        showResults: action.payload.showResults,
      };

    case SET_CREATED_POLL:
      return {
        ...state,
        created: action.payload.id,
      };

    case SET_OWNER:
      return {
        ...state,
        createdBy: action.payload.createdBy,
      };

    case RESET_POLL:
      return INITIAL_STATE;

    default:
      return state;
  }
};

export default combineReducers({ data: rootReducer, options: optionsReducer });
