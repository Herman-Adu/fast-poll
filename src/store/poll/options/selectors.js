import { createSelector } from 'reselect';

import { selectShowResults } from '../selectors';

export const selectOptionsOrder = state => state.poll.options.order;
export const selectOptions = state => state.poll.options.data;

export const selectOrderedOptions = createSelector(
  [selectOptionsOrder, selectOptions, selectShowResults],
  (order, options, showResults) => {
    if (showResults) {
      return Object.values(options).sort((a, b) => b.votes - a.votes);
    }

    return order.map(id => options[id]);
  },
);

export const selectTotalVotes = createSelector([selectOptions], options => {
  return Object.values(options).reduce((aggr, curr) => aggr + curr.votes, 0);
});
