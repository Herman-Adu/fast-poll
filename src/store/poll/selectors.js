// import { createSelector } from 'reselect';

import { selectUid } from '../auth/selectors';

export const selectCreatedBy = state => state.poll.data.createdBy;
export const selectShowResults = state => state.poll.data.showResults;

export const selectIsOwner = state => {
  return selectCreatedBy(state) === selectUid(state);
};
