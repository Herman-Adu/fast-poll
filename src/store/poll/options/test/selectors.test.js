import { INITIAL_STATE } from '../reducer';
import * as selectors from '../selectors';

describe('poll/options selectors', () => {
  const getState = showResults => {
    return {
      poll: {
        data: {
          showResults,
        },
        options: {
          ...INITIAL_STATE,
          order: ['1', '2', '3'],
          data: {
            '1': {
              id: '1',
              text: '1 poll',
              editing: false,
              votes: 2,
            },
            '3': {
              id: '3',
              text: '',
              editing: false,
              votes: 0,
            },
            '2': {
              id: '2',
              text: '2 poll',
              editing: false,
              votes: 1,
            },
          },
        },
      },
    };
  };

  describe('selectOrderedOptions', () => {
    it(`selects and sorts options while editing`, () => {
      expect(selectors.selectOrderedOptions(getState(false))).toMatchSnapshot();
    });

    it(`selects and sorts options after voted`, () => {
      expect(selectors.selectOrderedOptions(getState(true))).toMatchSnapshot();
    });
  });
});
