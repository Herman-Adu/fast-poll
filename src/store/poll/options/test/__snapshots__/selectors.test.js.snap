// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`poll/options selectors selectOrderedOptions selects and sorts options after voted 1`] = `
Array [
  Object {
    "editing": false,
    "id": "1",
    "text": "1 poll",
    "votes": 2,
  },
  Object {
    "editing": false,
    "id": "2",
    "text": "2 poll",
    "votes": 1,
  },
  Object {
    "editing": false,
    "id": "3",
    "text": "",
    "votes": 0,
  },
]
`;

exports[`poll/options selectors selectOrderedOptions selects and sorts options while editing 1`] = `
Array [
  Object {
    "editing": false,
    "id": "1",
    "text": "1 poll",
    "votes": 2,
  },
  Object {
    "editing": false,
    "id": "2",
    "text": "2 poll",
    "votes": 1,
  },
  Object {
    "editing": false,
    "id": "3",
    "text": "",
    "votes": 0,
  },
]
`;
